### PR TITLE
Switch to tokio-timer and lay the groundwork for mock timers in tests

### LINKED ISSUE #412: Replace timeout.rs with tower-timeout
[`proxy/timeout.rs`](https://github.com/runconduit/conduit/blob/master/proxy/src/timeout.rs) contains an implementation of a timeout Tower middleware. `tower` also contains a [`tower-timeout`](https://github.com/tower-rs/tower/tree/master/tower-timeout) crate that does pretty much the same thing. We may want to replace our hand-rolled timeout implementation with the middleware in `tower-timeout`.

### PR DESCRIPTION
**Pull request by hawkw**
This PR rewrites the `conduit_proxy::timeout` module to include a wrapper that abstracts over timer implementations, allowing us to transparently replace the default timer with a mock timer in tests. It also configures everything that used the old timeout code to use the new timers instead. The default timer is now based on `tokio_timer` rather than `tokio_core::reactor::Timeout`, which is likely more accurate and performant, and gives us a dedicated timer thread.

I had to increase the `DEFAULT_PRIVATE_CONNECT_TIMEOUT` value, as the old value was causing several tests to fail due to timeouts. I suspect the value was always too low, but the previous timer was not accurate enough for that to matter.

Closes #412. Unblocks #455.

### CODE REVIEW

**Comment by briansmith** [comment]
I thin kwe should verify that we can implement the mock timer in a reasonable way with this strategy before we spend time reviewing it. In particular, if we can't implement the mock timer using this strategy then we can skip over this to find an alternative that we can mock.

**Comment by hawkw** [comment]
@briansmith I have a fairly clear plan in my head for how to implement a mock timer, but haven't actually done it yet as other work is higher priority. I figured I'd put this up for review now anyway, but if you prefer I'd be happy to wait on merging it until there's a working mock timer implementation as well?

**Comment by briansmith** [comment]
OK. If I have time to review this after making more progress on my 0.3.1-prioritized work I'll take a look at it. I think the primary advantage of this change is improving the testing b/c this is where we can add the mocks, so if you can make progress on the mocks without a review here then that sounds good to me.

File: proxy/src/config.rs [hunk 1]
```diff
@@ -158,7 +158,7 @@ const DEFAULT_REPORT_TIMEOUT_SECS: u64 = 10; // TODO: is this a reasonable defau
 const DEFAULT_PRIVATE_LISTENER: &str = "tcp://127.0.0.1:4140";
 const DEFAULT_PUBLIC_LISTENER: &str = "tcp://0.0.0.0:4143";
 const DEFAULT_CONTROL_LISTENER: &str = "tcp://0.0.0.0:4190";
-const DEFAULT_PRIVATE_CONNECT_TIMEOUT_MS: u64 = 20;
+const DEFAULT_PRIVATE_CONNECT_TIMEOUT_MS: u64 = 200;
```
Comment (briansmith) [inline proxy/src/config.rs]: Please pull this change out as a separate PR and let's land that PR right away to see if it might help with #478.

I doubt that 200ms is a long enough timeout. Why not 10 seconds like the other timeouts?

Comment (hawkw) [reply proxy/src/config.rs]: Okay, I'd be happy to open a separate PR for this change.

200 ms was chosen fairly arbitrarily. I was assuming there was a reason for the original value being low and didn't want to increase the timeout too far beyond what was necessary. 50 and 100 ms were too low to fix the tests, at least on my box.

Comment (briansmith) [reply proxy/src/config.rs]: I think 10 seconds is a much better timeout than what we have now. Imagine that the service is implemented in a GC'd language. Note here https://stackoverflow.com/questions/574021/reducing-jvm-pause-time-1-second-using-useconcmarksweepgc that somebody is asking how to reduce a GC pause of over 1 second. There's no way 200 ms will be sufficient.

Anyway, let's discuss it further in its own PR.

Comment (hawkw) [reply proxy/src/config.rs]: 10s seems completely reasonable to me.

**Comment by briansmith** [comment]
> The default timer is now based on tokio_timer rather than tokio_core::reactor::Timeout, which is likely more accurate and performant, and gives us a dedicated timer thread.

/cc @carllerche 

One question I have regarding this: Do we really want/need to have a separate timer thread? The Conduit proxy is already adding threads (in a separate process) to each pod, which means it is increasing threading-related contention for the other containers in the pod. Further, there could be multiple Conduit proxies on the same node, when there are multiple pods running on the node that each have Conduit injected into them. We're likely making a significant contribution to the node running with more threads than cores, creating contention. Therefore, it seems good for us, in general, to minimize the number of threads we create in the proxy.

Given that, what is the specific motivation for having a dedicated timer thread, besides cleaner code? (I think we could probably get cleaner code without a dedicated timer thread.) Is there no reasonable way to get good performance without the separate thread?

**Comment by carllerche** [comment]
> which is likely more accurate and performant, and gives us a dedicated timer thread.

It is most likely more performant, but definitely less accurate (by design).

@briansmith The next iteration of tokio-timer will not require a dedicated thread.

**Comment by hawkw** [comment]
@briansmith et al, just a quick update on this: I have a separate branch which rewrites all the existing timing-dependent tests to use the mock timer instead. It still depends on `tokio-timer`. 

It's trivial to implement the timer facade for a reactor handle as well, which I've already done on the branch, but actually wiring that up is trickier. We'd want `Main::new` to be able to take a timer parameter for injecting the mock timer in tests, but in a case where we're using a `Handle` as the timer, we can't do this until the `Core` is actually created. I'm sure I could figure out a solution, though, given a bit more time.

Should I go ahead and open a PR for the current set of changes, relying on `tokio-timer`, or should I wait until we can make the mock timer change without switching our production timer from the core timer to `tokio-timer`?

**Comment by briansmith** [comment]
> Should I go ahead and open a PR for the current set of changes, relying on tokio-timer, or should I wait until we can make the mock timer change without switching our production timer from the core timer to tokio-timer?

I don't think we want to use the current version of tokio-timer that adds another background thread, especially if even the tokio-timer developers don't think that having another background thread is a good idea.

**Comment by hawkw** [comment]
> I don't think we want to use the current version of tokio-timer that adds another background thread, especially if even the tokio-timer developers don't think that having another background thread is a good idea.

Okay. Will hold off on opening a PR until I have it working with the reactor timer, then, thanks!

**Comment by hawkw** [comment]
I've opened #534, which adds a mock timer and rewrites the tests to use it. It also removes the `tokio-timer` dependency introduced in this branch and goes back to using reactor timeouts as we do currently.
